### PR TITLE
correcting the sign of the accumulation term of StandardWell

### PR DESCRIPTION
**Pull request by GitPaean**
Following the sign of the production rates and injection rates.

It is something we figured out when improving the multisegment well model, while it impacts much more cases with StandardWell, which are the most of the cases. 

So we created this PR for StandardWell separately for testing.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
It is great if someone has the privilege to fire a benchmarks to show the difference in term of performance.

My local testing did not show much difference, it is understandable for StandardWell model.

**Comment by alfbr** [comment]
You have the privilege :-) Just write benchmark

**Comment by GitPaean** [comment]
benchmark

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=101 please

**Comment by GitPaean** [comment]
benchmark

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.96     |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.995    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.04     |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.968    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=277

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.997    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.962    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.012    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.976    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=279

**Comment by GitPaean** [comment]
Norne with 1 process slightly faster while the other three setups are slightly slower.  For this situation, we should stick with the (more) correct formulation.

File: opm/autodiff/StandardWell_impl.hpp [hunk 1]
```diff
@@ -632,7 +632,9 @@ namespace Opm
 
         // add vol * dF/dt + Q to the well equations;
         for (int componentIdx = 0; componentIdx < numWellConservationEq; ++componentIdx) {
-            EvalWell resWell_loc = (wellSurfaceVolumeFraction(componentIdx) - F0_[componentIdx]) * volume / dt;
+            // TODO: following the development in MSW, we need to convert the volume of the wellbore to be surface volume
+            // since all the rates are under surface condition
+            EvalWell resWell_loc = (F0_[componentIdx] - wellSurfaceVolumeFraction(componentIdx)) * volume / dt;
             resWell_loc += getQs(componentIdx) * well_efficiency_factor_;
```
Comment (totto82) [inline opm/autodiff/StandardWell_impl.hpp]: Thanks for the fix.

File: opm/autodiff/StandardWell_impl.hpp [hunk 2]
```diff
@@ -2062,6 +2064,8 @@ namespace Opm
     calculateExplicitQuantities(const Simulator& ebosSimulator,
                                 const WellState& well_state)
     {
+        updatePrimaryVariables(well_state);
+        initPrimaryVariablesEvaluation();
         computeWellConnectionPressures(ebosSimulator, well_state);
```
Comment (totto82) [inline opm/autodiff/StandardWell_impl.hpp]: I guess this is an unrelated fix. Is this just to be on the safe side or did the lack of these line cause any problems?

File: opm/autodiff/StandardWell_impl.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: With some discussion with @atgeirr and @hnil , we think it is good to make the code in the form of (A-A0)/dt for ease of reading, which means we need to change the sign in other places. 

The current PR is just the way to do it with the least code change. 

I will revise this PR following the discussion.

File: opm/autodiff/StandardWell_impl.hpp (hunk 2 above)
Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: Basically, when all the primary variables are not initialized, we get all oil in the wellbore. 

Q_t  and F_w and F_g are all zero, we get F_o = 1. It is what happened in the simulation. It is a fix. I did not put it in another commit, only mentioned it in the commit message. 

I will split it to two commits.

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.993    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.978    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.039    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.97     |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=280

File: opm/autodiff/StandardWell_impl.hpp (hunk 2 above)
Comment (totto82) [reply opm/autodiff/StandardWell_impl.hpp]: You don't need to split it IMO, if you don't want to.

Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: okay. I need to revise this PR following the discussion to make the formulation easier to understand for later. We will have new Jenkins test and benchmark.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=101 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=101 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=101 please

**Comment by GitPaean** [comment]
It is not easy to maintain two StandardWell. We need a solution for this. Being able to incorporate the dynamic number of primary variables and equations is a great feature, while we need to do something to reduce the performance overhead.

**Comment by GitPaean** [comment]
benchmark

**Comment by GitPaean** [comment]
It looks like the jenkins test failed in an unexpected manner. I did not figure out the cause. 

Compared with the original version, I just flipped the sign of all the items in the well conservation equations.

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.993    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.907    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.028    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 1.021    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=281

**Comment by atgeirr** [comment]
> It is not easy and error-prone to maintain two StandardWell. We need a solution for this.

We should refactor this as soon as we finish the multisegment well improvements.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=101 please

**Comment by GitPaean** [comment]
It is fixed. I should not have flipped the sign for the matrix C. However, it looks like the benchmark performance does not looks bad at all. 

It caused big problem for the cases in opm-tests.

**Comment by GitPaean** [comment]
benchmark

**Comment by atgeirr** [comment]
Thanks, merging!

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.064    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.067    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.067    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.067    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=283

**Comment by GitPaean** [comment]
I think the above one is the bench mark results when this PR was not fixed.  It caused huge convergence problem and probably occupied the server for quite some hours.  Maybe the server side should have some time limits about how long a simulation can run.
